HypeBears.sol

```diff
@@ -0,0 +1,182 @@
+        whitelist[_newAddress] = true;
+    }
+
+    function removeFromWhitelist(address _address) external onlyOwner {
+        whitelist[_address] = false;
+    }
+
+    function addMultipleToWhitelist(address[] calldata _addresses) external onlyOwner {
+        require(_addresses.length <= 10000, "Provide less addresses in one function call");
+        for (uint256 i = 0; i < _addresses.length; i++) {
+            whitelist[_addresses[i]] = true;
+        }
+    }
+
+    function removeMultipleFromWhitelist(address[] calldata _addresses) external onlyOwner {
+        require(_addresses.length <= 10000, "Provide less addresses in one function call");
+        for (uint256 i = 0; i < _addresses.length; i++) {
+            whitelist[_addresses[i]] = false;
+        }
+    }
+
+    function canMint(address _address, bytes memory _signature) public view returns (bool, string memory) {
+        if (!whitelist[_address]) {
+            bytes32 hash = keccak256(abi.encodePacked(whitelistSigner, _address));
+            bytes32 messageHash = hash.toEthSignedMessageHash();
+
+            address signer = messageHash.recover(_signature);
+
+            if (signer != whitelistSigner) {
+                return (false, "Invalid signature");
+            }
+        }
+
+        if (addressMinted[_address]) {
+            return (false, "Already withdrawn");
+        }
+        return (true, "");
+    }
+
+    function withdraw() public onlyOwner {
+        uint256 balance = address(this).balance;
+        uint256 amount1 = balance * 70 / 100;
+        uint256 amount2 = balance - amount1;
+        payable(0xe0F7204f04b060715f858Ba8Ae357f57E5494d18).transfer(amount1);
+        payable(0x029c2D9EDC080A5A077f30F3bf6122e100F2aDc6).transfer(amount2);
+    }
+
+    function updateMintLimit(uint256 _newLimit) public onlyOwner {
+        mintLimit = _newLimit;
+    }
+
+    function updateMintPrice(uint256 _newPrice) public onlyOwner {
+        mintPrice = _newPrice;
+    }
+
+    function addPartnerMint(address account, uint256 amount) public onlyOwner {
+        partnerMintAmount += amount;
+        require(totalSupply().add(partnerMintAmount) <= TOTAL_NFT, "Can't add partner more than available");
+        partnerMintAvailableBy[account] += amount;
+    }
+
+    function mintNFT(uint256 _numOfTokens, bytes memory _signature) public payable {
+        require(mintActive, 'Not active');
+        require(_numOfTokens <= mintLimit, "Can't mint more than limit per tx");
+        require(mintPrice.mul(_numOfTokens) <= msg.value, "Insufficient payable value");
+        require(totalSupply().add(_numOfTokens).add(partnerMintAmount) <= TOTAL_NFT, "Can't mint more than 10000");
+        (bool success, string memory reason) = canMint(msg.sender, _signature);
+        require(success, reason);
+
+        for(uint i = 0; i < _numOfTokens; i++) {
+            _safeMint(msg.sender, totalSupply() + 1);
+        }
```

Comment:
Performing expensive operations in a loop, I think we can do better these days. More details around the inefficiencies at https://www.erc721a.org/

I highly suggest projects to look into ERC721A to understand how the various optimizations improve the performance, and either figure out how they can do something similar, or just get the optimizations for free with ERC721A